25/add inventory filtering

AdamVig opened the pull request:
Closes #25.

This also paves the way for adding filtering for rentals.

AdamVig committed:
Simplify sending response in endpoint getAll()

AdamVig committed:
Add request param to endpoint modifier functions

This will allow modifier functions to access data passed in the request
while modifying the database query. This could be useful specifically in
filtering, where specific fields and values from the request must be
used to add clauses to the database query.

AdamVig committed:
Add generic query filtering service

This service can be used inside a query modifier in an endpoint
controller to add filters to a query based on request parameters.

AdamVig committed:
Add inventory filtering

This allows request parameters to specify filters for the get all items
endpoint.

AdamVig committed:
Add inventory filtering documentation

emroussel commented on controllers/item.js:
```diff
@@ -19,11 +19,23 @@ item.withFields = (queryBuilder) => {
+    brand: 'brand.brandID',
+    model: 'model.modelID',
+    category: 'category.categoryID'
+  }
```
Would it make sense to name the params `brandID`, `modelID`, etc instead since we have to send the ids?

AdamVig replied:
I was thinking that, too. Would you also want the ability to send the names of brands, models, and categories?

emroussel replied:
I don't think that'd be necessary. I'll always have the ids since I have to get the brands, models and categories to fill the filtering options anyway. And comparing with ids seem better than doing it with strings. So I don't think it is worth implementing for now at least.

AdamVig committed:
Add "ID" to inventory filter parameter names

This is clearer because the parameters expect their values to be IDs.

AdamVig committed:
Change hash to map to fix for..in problem

When using an object hash as a map, a for..in loop needs to check that
each property is actually defined on the object itself. To avoid that,
an ES6 Map can be used. This has the side effect of allowing the use of
a for..of loop, which is cleaner than the for..in loop.
